src/addons/Portal/Portal.js

```diff
@@ -0,0 +1,255 @@
+    if (prevState.open && !this.state.open) {
+      debug('portal closed')
+      this.unmountPortal()
+    }
+  }
+
+  componentWillUnmount() {
+    this.unmountPortal()
+  }
+
+  // ----------------------------------------
+  // Document Event Handlers
+  // ----------------------------------------
+
+  closeOnDocumentClick = (e) => {
+    if (!this.props.closeOnDocumentClick) return
+
+    debug('closeOnDocumentClick()')
+
+    e.stopPropagation()
+    this.close()
+  }
+
+  closeOnEscape = (e) => {
+    if (!this.props.closeOnEscape) return
+    if (keyboardKey.getCode(e) !== keyboardKey.Escape) return
+
+    debug('closeOnEscape()')
+
+    e.preventDefault()
+    this.close()
+  }
+
+  // ----------------------------------------
+  // Component Event Handlers
+  // ----------------------------------------
+
+  handleChildClick = (e) => {
+    debug('handleChildClick()')
+
+    // Prevent closeOnDocumentClick
+    e.nativeEvent.stopImmediatePropagation()
+  }
+
+  handleTriggerBlur = (e) => {
+    if (!this.props.closeOnTriggerBlur) return
+
+    debug('handleTriggerBlur()')
+    this.close()
+  }
+
+  handleTriggerClick = (e) => {
+    if (!this.props.openOnTriggerClick) return
+
+    debug('handleTriggerClick()')
+
+    e.stopPropagation()
+    this.open()
+  }
+
+  handleTriggerFocus = (e) => {
+    if (!this.props.openOnTriggerFocus) return
+
+    debug('handleTriggerFocus()')
+    this.open()
+  }
+
+  handleTriggerMouseLeave = (e) => {
+    if (!this.props.closeOnTriggerMouseLeave) return
+
+    debug('handleTriggerMouseLeave()')
+    this.close()
+  }
+
+  handleTriggerMouseOver = (e) => {
+    if (!this.props.openOnTriggerMouseOver) return
+
+    debug('handleTriggerMouseOver()')
+    this.open()
+  }
+
+  // ----------------------------------------
+  // Behavior
+  // ----------------------------------------
+
+  open = () => {
+    debug('open()')
+    this.trySetState({ open: true })
+  }
+
+  close = () => {
+    debug('close()')
+    this.trySetState({ open: false })
+  }
+
+  renderPortal() {
+    const { children, className } = this.props
+
+    this.mountPortal()
+
+    this.node.className = className
+
+    const child = cloneElement(Children.only(children), {
+      onClick: this.handleChildClick,
+    })
+
+    this.portal = ReactDOM.unstable_renderSubtreeIntoContainer(
+      this,
+      child,
+      this.node
+    )
+  }
+
+  mountPortal = () => {
+    if (this.node) return
+
+    this.node = document.createElement('div')
+    document.body.appendChild(this.node)
+
+    const { onOpen } = this.props
+    if (onOpen) onOpen()
+
+    document.addEventListener('keydown', this.closeOnEscape)
+    document.addEventListener('click', this.closeOnDocumentClick)
+  }
+
+  unmountPortal = () => {
+    if (!this.node) return
+
+    ReactDOM.unmountComponentAtNode(this.node)
+    document.body.removeChild(this.node)
+
+    this.node = null
+    this.portal = null
+
+    const { onClose } = this.props
+    if (onClose) onClose()
```

Comment:
Yes sir, this another reason to only use user events to trigger handlers.  Programmatic changes do not have reliable events to pass back.

We should call the corresponding `trigger` event handler in `handleTriggerClick` `handleTriggerFocus` `handleTriggerMouseLeave` and `handleTriggerMouseOver` and pass the event.  Noted above in [this comment](https://github.com/TechnologyAdvice/stardust/pull/571/files#r81428230).

We should pass the event from each of those to the `open`/`close` method as well so that it can be passed to the `onOpen` and `onClose` callbacks.  This way, all callbacks get all the events possible.